### PR TITLE
✨  [Story video] Add quality sources on amp-video

### LINKED ISSUE #32337: [Story video] Add cached sources with different qualities to amp-video
Create sources with different bitrates for amp-video cached sources.

```html
<amp-video>
  <source src="cache?amp_quality=high" data-bitrate="hardcoded_value">
  <source src="cache?amp_quality=medium" data-bitrate="hardcoded_value">
  <source src="cache?amp_quality=low" data-bitrate="hardcoded_value">
  <source src="origin">
</amp-video>
```

Pointers:
https://github.com/ampproject/amphtml/blob/855a2a3d74b3df27e6b99d57fc26a11a144b985a/extensions/amp-video/0.1/amp-video.js#L438-L450

https://github.com/ampproject/amphtml/blob/855a2a3d74b3df27e6b99d57fc26a11a144b985a/extensions/amp-video/0.1/amp-video.js#L126-L160

### PR DESCRIPTION
**Pull request by mszylkowski**
Closes #32337

Adds 3 video qualities for ABR (adaptive bit rate):
 - High: 2000 kbps
 - Medium: 720 kbps
 - Low: 400 kbps

When the source of the video is cached, create the 3 sources with the different qualities and the ABR algorithm will use them.

Output:
```html
<video>
  <source src="./cachedurl.mp4?amp_quality=high" type="video/mp4" data-bitrate="2000">
  <source src="./cachedurl.mp4?amp_quality=medium" type="video/mp4" data-bitrate="720">
  <source src="./cachedurl.mp4?amp_quality=low" amp-orig-src="./originurl.mp4" type="video/mp4" data-bitrate="400">
  <source src="./originurl.mp4" type="video/mp4">
</video>
```

### CODE REVIEW

File: extensions/amp-video/0.1/test/test-amp-video.js [hunk 1]
```diff
@@ -163,15 +169,11 @@ describes.realWin(
...
-      expect(sources[1].getAttribute('src')).to.equal(
-        'https://example.com/video.mp4'
-      );
-      expect(sources[0].getAttribute('data-bitrate')).to.equal('1000');
-      expect(sources[1].getAttribute('data-bitrate')).to.equal('1000');
+      expect(sources.length).to.equal(4);
+      expect(sources[0].getAttribute('data-bitrate')).to.equal('2000');
+      expect(sources[1].getAttribute('data-bitrate')).to.equal('720');
+      expect(sources[2].getAttribute('data-bitrate')).to.equal('400');
+      expect(sources[3].hasAttribute('data-bitrate')).to.be.false;
```
Comment (mszylkowski) [inline extensions/amp-video/0.1/test/test-amp-video.js:176]: If the source comes with a bitrate should we list that bitrate here or remove it and rely on the other qualities?

Comment (gmajoulet) [reply extensions/amp-video/0.1/test/test-amp-video.js:176]: Good Q.
If the publisher provided a bitrate themselves, we should keep it. One implementation could be:
- When creating the last source with `amp_quality=...`, if the publisher provided a bitrate themselves, move `data-bitrate` to `data-amp-orig-bitrate`
- Line 502 where you removed the bitrate attr propagation, reintroduce the code but propagate the `data-amp-orig-bitrate`

wdyt?

File: extensions/amp-video/0.1/amp-video.js [hunk 2]
```diff
@@ -63,6 +64,13 @@ const ATTRS_TO_PROPAGATE_ON_BUILD = [
   'controlsList',
 ];
 
+/** @private {!Map<string, number>} the bitrate in Kb/s of amp_quality */
```
[the remark this answers is not in the document]
Comment (mszylkowski) [reply extensions/amp-video/0.1/amp-video.js:67]: Discussed offline we're keeping it as enum, closing this comment

File: extensions/amp-video/0.1/amp-video.js [hunk 3]
```diff
@@ -454,11 +462,30 @@ class AmpVideo extends AMP.BaseElement {
...
+        qualities.forEach((quality) => {
+          const cachedSource = addParamsToUrl(source.src, {
+            'amp_quality': quality,
+          });
+          const currSource = this.createSourceElement_(
+            cachedSource,
+            origType,
+            AMP_QUALITY_BITRATES[quality]
+          );
+          if (qualities.indexOf(quality) == qualities.length - 1) {
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/amp-video.js:482]: - Please use the `index` param of the forEach callback
- Use strict equality operator

File: extensions/amp-video/0.1/amp-video.js [hunk 4]
```diff
@@ -454,11 +462,30 @@ class AmpVideo extends AMP.BaseElement {
...
+            // Keep src of amp-orig only in last one so it adds the orig source after it.
+            // TODO: Propagate data-bitrate of source.
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/amp-video.js:484]: Please remove this TODO as it doens't have any context for anyone to act on it

File: extensions/amp-video/0.1/amp-video.js [hunk 5]
```diff
@@ -534,17 +557,21 @@ class AmpVideo extends AMP.BaseElement {
   /**
    * @param {string} src
    * @param {?string} type
+   * @param {?string} bitrate
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/amp-video.js:560]: if this parameter is optional: `string=`
But from your map it's a `number`?

File: extensions/amp-video/0.1/amp-video.js [hunk 6]
```diff
@@ -534,17 +557,21 @@ class AmpVideo extends AMP.BaseElement {
...
-  createSourceElement_(src, type) {
+  createSourceElement_(src, type, bitrate = null) {
     const {element} = this;
     this.getUrlService_().assertHttpsUrl(src, element);
     const source = element.ownerDocument.createElement('source');
     source.setAttribute('src', src);
     if (type) {
       source.setAttribute('type', type);
     }
+    if (bitrate) {
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/amp-video.js:572]: Nit: `bitrate !== null` to allow for 0

File: extensions/amp-video/0.1/amp-video.js (hunk 5 above)
Comment (mszylkowski) [reply extensions/amp-video/0.1/amp-video.js:560]: `string|number=`?

File: extensions/amp-video/0.1/amp-video.js [hunk 7]
```diff
@@ -534,17 +559,21 @@ class AmpVideo extends AMP.BaseElement {
   /**
    * @param {string} src
    * @param {?string} type
+   * @param {string|number=} bitrate
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/amp-video.js:562]: I don't think you need the `string` here?

**Review by gmajoulet — APPROVED** [review]

**Pull request merged**